programs/whirlpool/src/util/v2/token.rs

```diff
@@ -0,0 +1,347 @@
+    if let Ok(memo_transfer) = extension {
+        return Ok(memo_transfer.require_incoming_transfer_memos.into());
+    } else {
+        return Ok(false);
+    }
+}
+
+pub fn is_supported_token_mint<'info>(
+    token_mint: &InterfaceAccount<'info, Mint>,
+    is_token_badge_initialized: bool,
+) -> Result<bool> {
+    let token_mint_info = token_mint.to_account_info();
+
+    // if mint is owned by Token Program, it is supported (compatible to initialize_pool / initialize_reward)
+    if *token_mint_info.owner == Token::id() {
+        return Ok(true);
+    }
+
+    // now mint is owned by Token-2022 Program
+
+    // reject if mint has freeze_authority
+    if token_mint.freeze_authority.is_some() && !is_token_badge_initialized {
+        return Ok(false);
+    }
+
+    let token_mint_data = token_mint_info.try_borrow_data()?;
+    let token_mint_unpacked = StateWithExtensions::<spl_token_2022::state::Mint>::unpack(&token_mint_data)?;
+
+    let extensions = token_mint_unpacked.get_extension_types()?;
+    for extension in extensions {
+        match extension {
+            // supported
+            extension::ExtensionType::TransferFeeConfig => {}
+            extension::ExtensionType::TokenMetadata => {}
+            extension::ExtensionType::MetadataPointer => {}
+            // partially supported
+            extension::ExtensionType::ConfidentialTransferMint => {
+                // Supported, but non-confidential transfer only
+                //
+                // WhirlpoolProgram invokes TransferChecked instruction and it supports non-confidential transfer only.
+                //
+                // Because the vault accounts are not configured to support confidential transfer,
+                // it is impossible to send tokens directly to the vault accounts confidentially.
+                // Note: Only the owner (Whirlpool account) can call ConfidentialTransferInstruction::ConfigureAccount.
+            }
+            // supported if token badge is initialized
+            extension::ExtensionType::PermanentDelegate => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::TransferHook => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::MintCloseAuthority => {
+                if !is_token_badge_initialized { return Ok(false); }
+            }
+            extension::ExtensionType::DefaultAccountState => {
+                if !is_token_badge_initialized { return Ok(false); }
+
+                // reject if default state is not Initialized even if it has token badge
+                let default_state = token_mint_unpacked.get_extension::<extension::default_account_state::DefaultAccountState>()?;
+                let initialized: u8 = AccountState::Initialized.into();
+                if default_state.state != initialized {
+                    return Ok(false);
+                }
```

Comment:
I believe `TokenBadge` tokens are "trusted" with freeze authority. In this case, perhaps we don't need this check?
TokenBadge Mint owners can bypass this check by changing the default state away from the unsupported states here anyway

Reply:
FreezeAuthority and PermanentDelegate are acceptable with TokenBadge because stable coins will use them. However, that does not necessarily mean that Orca welcome freezing the Whirlpool vault.
Pools with frozen vault can cause dysfunction and negatively impact the ecosystem.

This function is called when initialize_pool_v2 is executed. That is, if the initial value is Frozen at this point, the pool will be created with vault frozen. The intent is to not allow the creation of a dysfunctional pool from the initial state. 
Of course, it does not prevent the pool from being frozen within the transaction that creates it. If this behavior is observed, I think TokenBadge should be revoked. WDYT ?